File: src/sugtao4423/lod/tweetlistview/EndlessScrollListener.java
package sugtao4423.lod.tweetlistview;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

public abstract class EndlessScrollListener extends RecyclerView.OnScrollListener{

	int visibleThreshold = 5;
	int visibleItemCount, totalItemCount;
	private int previousTotal = 0;
	private boolean loading = true;
	private int current_page = 0;

	private LinearLayoutManager mLinearLayoutManager;

	public EndlessScrollListener(LinearLayoutManager linearLayoutManager){
		this.mLinearLayoutManager = linearLayoutManager;
	}

	public void setVisibleThreshold(int visibleThreshold){
		this.visibleThreshold = visibleThreshold;
	}

	@Override
	public void onScrolled(RecyclerView recyclerView, int dx, int dy){
		super.onScrolled(recyclerView, dx, dy);

		visibleItemCount = recyclerView.getChildCount();
		totalItemCount = mLinearLayoutManager.getItemCount();
		int lastVisibleItem = mLinearLayoutManager.findLastVisibleItemPosition();

		if(loading && totalItemCount > previousTotal){
			loading = false;
			previousTotal = totalItemCount;
		}

		if(!loading && (lastVisibleItem + visibleThreshold) > totalItemCount){
			current_page++;
			onLoadMore(current_page);
			loading = true;
		}
	}

	public void resetState() {
        this.current_page = 0;
        this.previousTotal = 0;
        this.loading = true;
    }

	public abstract void onLoadMore(int current_page);
}

File: app/src/main/java/sugtao4423/lod/tweetlistview/EndlessScrollListener.kt
package sugtao4423.lod.tweetlistview

import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

abstract class EndlessScrollListener(private val linearLayoutManager: LinearLayoutManager) : RecyclerView.OnScrollListener() {

    var visibleThreshold = 5
    var visibleItemCount = -1
    var totalItemCount = -1
    private var previousTotal = 0
    private var loading = true
    private var currentPage = 0

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)

        visibleItemCount = recyclerView.childCount
        totalItemCount = linearLayoutManager.itemCount
        val lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition()

        if (loading && totalItemCount > previousTotal) {
            loading = false
            previousTotal = totalItemCount
        }

        if (!loading && (lastVisibleItem + visibleThreshold) > totalItemCount) {
            currentPage++
            onLoadMore(currentPage)
            loading = true
        }
    }

    fun resetState() {
        this.currentPage = 0
        this.previousTotal = 0
        this.loading = true
    }

    abstract fun onLoadMore(currentPage: Int)

}